validator/src/journal/chain.rs

```diff
@@ -0,0 +1,683 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+use std::fs::File;
+use std::io;
+use std::io::prelude::*;
+use std::marker::Send;
+use std::marker::Sync;
+use std::path::PathBuf;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+use std::sync::mpsc;
+use std::sync::mpsc::channel;
+use std::sync::mpsc::Receiver;
+use std::sync::mpsc::RecvError;
+use std::sync::mpsc::Sender;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::sync::RwLock;
+use std::thread;
+use std::time::Duration;
+
+use protobuf;
+
+use batch::Batch;
+use journal;
+use journal::block_wrapper::BlockWrapper;
+use metrics;
+
+use proto::transaction_receipt::TransactionReceipt;
+use scheduler::TxnExecutionResult;
+
+const RECV_TIMEOUT_MILLIS: u64 = 100;
+
+lazy_static! {
+    static ref COLLECTOR: metrics::MetricsCollectorHandle =
+        metrics::get_collector("sawtooth_validator.chain");
+}
+
+#[derive(Debug)]
+pub enum ChainControllerError {
+    QueueRecvError(RecvError),
+    ChainIdError(io::Error),
+    ChainUpdateError(String),
+    BlockValidationError(ValidationError),
+    BrokenQueue,
+}
+
+impl From<RecvError> for ChainControllerError {
+    fn from(err: RecvError) -> Self {
+        ChainControllerError::QueueRecvError(err)
+    }
+}
+
+impl From<io::Error> for ChainControllerError {
+    fn from(err: io::Error) -> Self {
+        ChainControllerError::ChainIdError(err)
+    }
+}
+
+impl From<ValidationError> for ChainControllerError {
+    fn from(err: ValidationError) -> Self {
+        ChainControllerError::BlockValidationError(err)
+    }
+}
+
+pub trait ChainObserver: Send + Sync {
+    fn chain_update(&mut self, block: &BlockWrapper, receipts: &[&TransactionReceipt]);
+}
+
+pub trait ExternalLock: Send + Sync {
+    fn lock(&self) -> Box<ExternalLockGuard>;
+}
+
+pub trait ExternalLockGuard: Drop {
+    fn release(&self);
+}
+
+pub trait ChainHeadUpdateObserver: Send + Sync {
+    /// Called when the chain head has updated.
+    ///
+    /// Args:
+    ///     block: the new chain head
+    ///     committed_batches: all of the batches that have been committed
+    ///         on the given fork. This may be across multiple blocks.
+    ///     uncommitted_batches: all of the batches that have been uncommitted
+    ///         from the previous fork, if one was dropped.
+    fn on_chain_head_updated(
+        &mut self,
+        block: BlockWrapper,
+        committed_batches: Vec<Batch>,
+        uncommitted_batches: Vec<Batch>,
+    );
+}
+
+pub trait BlockCache: Send + Sync {
+    fn contains(&self, block_id: &str) -> bool;
+
+    fn put(&mut self, block: BlockWrapper);
+
+    fn get(&self, block_id: &str) -> Option<BlockWrapper>;
+}
+
+#[derive(Debug)]
+pub enum ChainReadError {
+    GeneralReadError(String),
```

Comment:
Error types do not have to be enums. If there is only one kind of error, you can use a struct, new type, or tuple struct instead:
```
// Using a tuple struct
pub struct ChainHeadError(String);
```

Reply:
An interesting point, but I would prefer to remain consistent in the code by using enums for now.